### PR TITLE
Markdown extensions and table of contents

### PR DESCRIPTION
**Pull request by WillMoggridge**
- Markdown extensions
- Anchor links
- Table of contents on documentation pages

This change moves direct view markdown parsing, into the view itself. The main reason is that it means the view can access the table of contents and display it on the page anywhere the template decides.

The TOC is added to the documentation page_type, and we will need to add this page_type to pages shortly. It needs a little more styling but the functionality works!

### CODE REVIEW

File: webapp/lib/markdown/markdown.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    markdown_parser = _markdown.Markdown(extensions=markdown_extensions)
+    parsed_markdown = markdown_parser.convert(markdown_content)
+    table_of_contents = markdown_parser.toc
+    if table_of_contents:
+        metadata['table_of_contents'] = table_of_contents
+    return markdown_parser, parsed_markdown, metadata
+
+
+def parse_markdown(markdown_content):
```
Comment (nottrobin) [inline webapp/lib/markdown/markdown.py]: Flatten thjis function

File: webapp/lib/markdown/markdown.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        # We don't need any slashes at the ends
+        path.strip('/')
+
+        template_paths = [
+            ''.join([template_root, path, '.md']),
+        ]
+        try:
+            template = loader.select_template(template_paths)
+            template_path = template.origin.name
+        except TemplateDoesNotExist:
```
Comment (nottrobin) [inline webapp/lib/markdown/markdown.py]: Maybe don't override the original exception?

File: webapp/loaders.py [hunk 3]
```diff
@@ -71,13 +52,24 @@ def find_template(self, name, dirs=None):
...
-                    template_name, template_dirs
-                )
-            except TemplateDoesNotExist:
-                pass
+            for template_name in template_names:
+                try:
+                    return loader.load_template_source(
+                        template_name, template_dirs
+                    )
+                except TemplateDoesNotExist:
```
Comment (nottrobin) [inline webapp/loaders.py:69]: Should we be passing over this error?

**Review by nottrobin — CHANGES_REQUESTED** [review]
A few suggestions

File: webapp/views.py [hunk 4]
```diff
@@ -33,50 +38,63 @@ def get_template_names(self):
...
-    def get_context_data(self, **kwargs):
-        request_path = self.kwargs['path']
-        template, template_path = self._find_template(request_path)
-        with open(template.origin.name, 'r') as f:
-            metadata = parse_frontmatter(f.read())
+    def _parse_markdown(self, path):
+        markdown, template_path = self._find_template_source(path)
+        # parser, parsed_markdown, metadata = get_markdown_with_parser(markdown)
+        parsed_markdown, metadata = parse_markdown(markdown)
+
```
Comment (nottrobin) [inline webapp/views.py]: Formatting: remove extra whitespace

File: webapp/views.py [hunk 5]
```diff
@@ -33,50 +38,63 @@ def get_template_names(self):
     def _get_base_template_name(self):
         return self.template_name or self.kwargs['template_name']
 
-    def _find_template(self, path):
+    def _find_markdown_loader(self):
+        loaders = Engine.get_default().template_loaders
+        for loader in loaders:
```
Comment (nottrobin) [inline webapp/views.py]: formatting: masks earlier `loader` declaration

File: webapp/views.py [hunk 6]
```diff
@@ -6,9 +6,13 @@
     RequestContext,
     TemplateDoesNotExist,
 )
+from django.template.engine import Engine
 from django.views.generic import TemplateView
 
 from webapp.lib.markdown import parse_frontmatter
```
Comment (nottrobin) [inline webapp/views.py]: formatting: imported but unused

File: webapp/views.py [hunk 7]
```diff
@@ -6,9 +6,13 @@
...
+# from webapp.lib.markdown import get_markdown_with_parser
```
Comment (nottrobin) [inline webapp/views.py]: formatting: please remove commented out code

File: webapp/loaders.py (hunk 3 above)
Comment (WillMoggridge) [reply webapp/loaders.py:69]: OK, I remember now after sleep!
We are looping over the loaders and requesting template names. It expects it to return this exception sometimes.

So, the first loader may return the exception, then the second one finds the template.

**Review by nottrobin — CHANGES_REQUESTED** [review]
Some formatting points

**Comment by WillMoggridge** [comment]
Updated

**Review by nottrobin — APPROVED** [review]

**Pull request merged**